From 06875367a29510b90803e3e553d0172679b06747 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Aleix=20Conchillo=20Flaqu=C3=A9?= <aleix@oblong.com>
Date: Fri, 7 Nov 2014 17:52:49 -0800
Subject: [PATCH] rtspconnection: also flush output stream when flushing watch

It is possible that the connection output stream has internal data that
also needs flushing. We need to do that as we are not planning to resend
the queued RTSP records and the output stream might be in a wrong state.

https://bugzilla.gnome.org/show_bug.cgi?id=739799
---
 gst-libs/gst/rtsp/gstrtspconnection.c |   25 +++++++++++++++++++++++++
 1 file changed, 25 insertions(+)

diff --git a/gst-libs/gst/rtsp/gstrtspconnection.c b/gst-libs/gst/rtsp/gstrtspconnection.c
index 23ba516..9707711 100644
--- a/gst-libs/gst/rtsp/gstrtspconnection.c
+++ b/gst-libs/gst/rtsp/gstrtspconnection.c
@@ -3858,11 +3858,15 @@ timeout:
  * and make sure gst_rtsp_watch_write_data() returns immediately with
  * #GST_RTSP_EINTR. And empty the queue.
  *
+ * It also flushes the connection output stream.
+ *
  * Since: 1.4
  */
 void
 gst_rtsp_watch_set_flushing (GstRTSPWatch * watch, gboolean flushing)
 {
+  GError *err = NULL;
+
   g_return_if_fail (watch != NULL);
 
   g_mutex_lock (&watch->mutex);
@@ -3871,6 +3875,27 @@ gst_rtsp_watch_set_flushing (GstRTSPWatch * watch, gboolean flushing)
   if (flushing == TRUE) {
     g_queue_foreach (watch->messages, (GFunc) gst_rtsp_rec_free, NULL);
     g_queue_clear (watch->messages);
+
+    if (watch->conn->output_stream) {
+      gboolean res;
+
+      res = g_output_stream_flush (watch->conn->output_stream,
+          watch->conn->cancellable, &err);
+
+      if (G_UNLIKELY (!res))
+        goto error;
+    }
   }
   g_mutex_unlock (&watch->mutex);
+
+  return;
+
+  /* ERRORS */
+error:
+  {
+    GST_DEBUG ("Unable to flush output stream: %s", err->message);
+    g_clear_error (&err);
+    g_mutex_unlock (&watch->mutex);
+    return;
+  }
 }
-- 
1.7.9.5

